framework/util/spirv_parsing_util.cpp

```diff
@@ -0,0 +1,476 @@
+
+    if (storage_class == spv::StorageClassPushConstant)
+    {
+        buffer_reference_info.push_constant_block = true;
+        return true;
+    }
+
+    for (const Instruction* insn : decorations_instructions_)
+    {
+        if (insn->operand(0) != variable_id)
+        {
+            continue;
+        }
+        if (insn->operand(1) == spv::DecorationDescriptorSet)
+        {
+            buffer_reference_info.set = insn->operand(2);
+        }
+        else if (insn->operand(1) == spv::DecorationBinding)
+        {
+            buffer_reference_info.binding = insn->operand(2);
+        }
+    }
+
+    if (storage_class == spv::StorageClassStorageBuffer || storage_class == spv::StorageClassUniform)
+    {
+        return true;
+    }
+    else
+    {
+        printf("Storage class %u not handled\n", storage_class);
+        return false;
+    }
+}
+
+bool SpirVParsingUtil::Parse(const uint32_t* const spirv_code, size_t spirv_num_bytes)
+{
+    if (spirv_code == nullptr)
+    {
+        return false;
+    }
+
+    definitions_.clear();
+    store_instructions_.clear();
+    decorations_instructions_.clear();
+    buffer_reference_map_.clear();
+
+    // use in combination with spirv-reflect
+    std::optional<SpvReflectShaderModule> spv_shader_module;
+
+    // spirv-header is 5 d-words
+    constexpr uint32_t spirv_header_size = 5;
+    const uint32_t*    spirv_ptr         = spirv_code;
+    const uint32_t*    spirv_end         = spirv_code + (spirv_num_bytes / sizeof(uint32_t));
+
+    // skip header
+    spirv_ptr += spirv_header_size;
+
+    std::vector<Instruction> instructions;
+
+    // First build up instructions object to make it easier to work with the SPIR-V
+    while (spirv_ptr != spirv_end)
+    {
+        Instruction& insn = instructions.emplace_back(spirv_ptr);
+        spirv_ptr += insn.length();
+    }
+    instructions.shrink_to_fit();
+
+    auto track_back_instruction = [this, &spv_shader_module, spirv_code, spirv_num_bytes](
+                                      const Instruction* object_insn) {
+        // keep track of access-chain
+        std::vector<uint32_t> access_indices;
+
+        // We are where a buffer-reference was accessed, now walk back to find where it came from
+        while (object_insn)
+        {
+            switch (object_insn->opcode())
+            {
+                case spv::OpConvertUToPtr:
+                case spv::OpCopyLogical:
+                case spv::OpLoad:
+                    object_insn = FindDef(object_insn->operand(0));
+                    break;
+                case spv::OpAccessChain:
+                {
+                    std::vector<uint32_t> indices;
+                    for (uint32_t i = 1; i < object_insn->num_operands(); ++i)
+                    {
+                        if (auto ins = FindDef(object_insn->operand(i)))
+                        {
+                            if (ins->opcode() == spv::OpConstant)
```

Comment:
so this is fine for 99% of cases... but for VVL we ended up creating this helper for many reason

```
// While simple, function name provides a more human readable description why Word(3) is used.
//
// The current various uses for constant values (OpAccessChain, OpTypeArray, LocalSize, etc) all have spec langauge making sure they
// are scalar ints. It is also not valid for any of these use cases to have a negative value. While it is valid SPIR-V to use 64-bit
// int, found writting test there is no way to create something valid that also calls this function. So until a use-case is found,
// we can safely assume returning a uint32_t is ok.
uint32_t Instruction::GetConstantValue() const {
    // This should be a OpConstant (not a OpSpecConstant), if this asserts then 2 things are happening
    // 1. This function is being used where we don't actually know it is a constant and is a bug in the validation layers
    // 2. The CreateFoldSpecConstantOpAndCompositePass didn't fully fold everything and is a bug in spirv-opt
    assert(Opcode() == spv::OpConstant);
    return Word(3);
}
```

.... for now, I would just go `if not a `OpConstant` exit cleanly and mark as a "edge case"

Reply:
thinking about this more, the main 2 things to worry about is

1. `OpSpecConstant` (Which I am not sure if fully legal here without trying to create it manually in SPIR-V, so might be not something to worry about)
2. `OpConstant` **can** be 64-bit, but have found almost never in the wild

again, I think you have this wrapped around safely enough... just dropping a lot of un-fun SPIR-V knowledge

Reply:
here I was also frequently seeing an `OpLoad` instead, which I just ignored because it still yielded useful info. 
for a statement like: 
```
layout(buffer_reference, scalar) readonly buffer VertexBufferPtr { packed_vertex_t v[]; };
layout(binding = 0, set = 0) readonly buffer Vertices { VertexBufferPtr vertex_buffers[]; };
...
Vertex v = unpack(vertex_buffers[draws[indices.mesh_draw_index].mesh_index].v[gl_VertexIndex]);
```
we get:
`buffer-reference: (Vertices) -> vertex_buffers (set: 0, binding: 0, buffer-offset: 0, array-stride: 8)`
-> a buffer with tightly packed references, all we want to know

Reply:
this just means the access is dynamic

so if I went `data[3]` its will be `OpConstant 3` but if I go `data[index]` it will be an `OpLoad` to the `index` variable as its dynamic and you don't know which index, but you know still where it started from which is what leads you to the descriptor set/binding